src/Aspire.Hosting/api/Aspire.Hosting.cs

```diff
@@ -2542,20 +3031,279 @@ public partial interface IDistributedApplicationLifecycleHook
+
+        public Microsoft.Extensions.Logging.ILogger Logger { get { throw null; } }
+
+        public ApplicationModel.DistributedApplicationModel Model { get { throw null; } }
+
+        public required PipelineContext PipelineContext { get { throw null; } init { } }
+
+        public required IReportingStep ReportingStep { get { throw null; } init { } }
+
+        public System.IServiceProvider Services { get { throw null; } }
+    }
+
+    [System.Diagnostics.CodeAnalysis.Experimental("ASPIREPIPELINES001", UrlFormat = "https://aka.ms/aspire/diagnostics/{0}")]
+    public static partial class PipelineStepExtensions
+    {
+        public static ApplicationModel.IResourceBuilder<T> WithPipelineConfiguration<T>(this ApplicationModel.IResourceBuilder<T> builder, System.Action<PipelineConfigurationContext> callback)
+            where T : ApplicationModel.IResource { throw null; }
+
+        public static ApplicationModel.IResourceBuilder<T> WithPipelineConfiguration<T>(this ApplicationModel.IResourceBuilder<T> builder, System.Func<PipelineConfigurationContext, System.Threading.Tasks.Task> callback)
+            where T : ApplicationModel.IResource { throw null; }
+
+        public static ApplicationModel.IResourceBuilder<T> WithPipelineStepFactory<T>(this ApplicationModel.IResourceBuilder<T> builder, System.Func<PipelineStepFactoryContext, PipelineStep> factory)
+            where T : ApplicationModel.IResource { throw null; }
+
+        public static ApplicationModel.IResourceBuilder<T> WithPipelineStepFactory<T>(this ApplicationModel.IResourceBuilder<T> builder, System.Func<PipelineStepFactoryContext, System.Collections.Generic.IEnumerable<PipelineStep>> factory)
+            where T : ApplicationModel.IResource { throw null; }
+
+        public static ApplicationModel.IResourceBuilder<T> WithPipelineStepFactory<T>(this ApplicationModel.IResourceBuilder<T> builder, System.Func<PipelineStepFactoryContext, System.Threading.Tasks.Task<PipelineStep>> factory)
+            where T : ApplicationModel.IResource { throw null; }
+
+        public static ApplicationModel.IResourceBuilder<T> WithPipelineStepFactory<T>(this ApplicationModel.IResourceBuilder<T> builder, System.Func<PipelineStepFactoryContext, System.Threading.Tasks.Task<System.Collections.Generic.IEnumerable<PipelineStep>>> factory)
+            where T : ApplicationModel.IResource { throw null; }
+    }
+
+    [System.Diagnostics.CodeAnalysis.Experimental("ASPIREPIPELINES001", UrlFormat = "https://aka.ms/aspire/diagnostics/{0}")]
+    public partial class PipelineStepFactoryContext
+    {
+        public required PipelineContext PipelineContext { get { throw null; } init { } }
+
+        public required ApplicationModel.IResource Resource { get { throw null; } init { } }
+    }
+
+    [System.Diagnostics.CodeAnalysis.Experimental("ASPIREPIPELINES001", UrlFormat = "https://aka.ms/aspire/diagnostics/{0}")]
+    public static partial class PipelineStepsExtensions
```

Comment:
Do we really need 2 classes? `PipelineStepExtensions` and `PipelineStepsExtensions`? 

cc @captainsafia @davidfowl

Reply:
It seems like one category of extension methods is about constructing steps and the other is about wiring up dependencies. We can probably rename as follows:

- `PipelineStepExtensions` => `PipelineStepFactoryExtensions`
- `PipelineStepsExtensions` => `PipelineStepDependenciesExtensions` (or something idk i kinda don't like that name)